Improve how Istio citadel decides which set of namespaces to work on [Issue #14828]

Monkeyanator opened the pull request:
---
name: Improve how Istio citadel decides which set of namespaces to work on [Issue #14828]
about: Removed explicit-opt-in flag and moved towards the deprecation of listened-namespaces flag. Added new convention which expects "citadel.istio.io/managed" label to be present on resources. 
---

[Relevant issue
](https://github.com/istio/istio/issues/14828)

[Relevant design document](https://docs.google.com/document/d/1FRO_Cso_Q7FbJ5-TIXt1kIDirEurWRIAIZN8glr-5vc/edit#heading=h.7ubp6he6anu6)

Users should be able to configure which namespaces Citadel will target for its secret controller loop, but the current configuraiton options are undocumented, ambiguous, and inflexible. This PR introduces a new designated label, `citadel.istio.io/managed`, which allows users to tag which namespace Citadel should operate on, in addition to a new flag, `namespaces-managed-default`, which determines the default behavior when the designated label is not found (i.e. to create secret for added or not to create secret for added `ServiceAccount`).

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastrcture

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -209,40 +210,40 @@ func GetSecretName(saName string) string {
+			case "false", "no", "n":
+				isEnabled = false
+			default: // use default value
 				break
```
i would rather remove default and return `isEnabled` directly.

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -209,40 +210,40 @@ func GetSecretName(saName string) string {
+			default: // use default value
 				break
 			}
+		} else { // use default value
```
This hanging else branch shd be removed.

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -209,40 +210,40 @@ func GetSecretName(saName string) string {
-			case "disabled", "disable", "false", "no", "n":
-				enabled = false
-			default: // leave default unchanged
+			case "true", "yes", "y":
```
i feel we can just support true/false to reduce the complexity and user surface area.

this will be a new label, so should be fine.

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -209,40 +210,40 @@ func GetSecretName(saName string) string {
-	if _, watched := sc.namespaces[obj.GetNamespace()]; watched || !sc.explicitOptIn {
+func (sc *SecretController) istioManagedObject(obj metav1.Object) bool {
+
+	// should be removed once listened namespaces flag phased out
```
add a todo(github-name) either yours or mine, either is fine :)

Monkeyanator replied:
since I'm not sure how long phasing out the "listened-namespaces" flag will take, I'll put your github name down for now

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -209,40 +210,40 @@ func GetSecretName(saName string) string {
 	ns, err := sc.core.Namespaces().Get(obj.GetNamespace(), metav1.GetOptions{})
-	if err != nil || ns == nil { // @todo handle errors? Unit tests mocks don't create NS, only secrets
-		return enabled
+	if err != nil || ns == nil { // if we can't retrieve namespace details, fall back on default value
```
do you know why ns can be nil?

incfly replied:
also log.Error should be good, given that we're running into this without your rbac acl.

Monkeyanator replied:
Hm I'm not sure on that, I'll look more into it but I think the check can be just whether `err != nil`

incfly commented:
Thanks for the work! 

No need to rush for fixing the comments. I think it's helpful to think through some corner case and our handling first, maybe need to supplement the design a bit, which is the goal of this PR :)

Monkeyanator commented:
Thanks for taking the time to go through this! Just a few questions:

> We need a client to consume updates when customer add/remove labels to existing namespaces, right?
> 
> The code currently only get triggered when service account is added or removed. It's possible only event of labeling on namespace happen, no service account events. Citadel, by definition of this design, should take actions on that.

Just to make sure I understand: is this to handle the case where a user re-labels a namespace that has existing ServiceAccounts in it (either from active -> inactive, or inactive -> active)? Should those ServiceAccounts have secrets retroactively created/deleted after their namespace label is toggled?

Other than making secrets for the service accounts, what other actions should Citadel take here?

> Also we haven't defined the behavior when we found existing key/cert on some service account not needed ( because of labeling). Should we let them be there? Or add some logging/metrics about orphaned key/cert? Will delete them result into some dangerous situation?

I would tend to say (more from a user perspective)  that Citadel should handle cleaning up these key/certs automatically in disabled namespaces if possible. However, I don't fully understand the implications of doing that, and where dangerous situations which could arise, so I'll add a section in the design doc about it, and ask for feedback on it.

incfly commented:
> Just to make sure I understand: is this to handle the case where a user re-labels a namespace that has existing ServiceAccounts in it (either from active -> inactive, or inactive -> active)? Should those ServiceAccounts have secrets retroactively created/deleted after their namespace label is toggled?

Yes, your understanding is correct. Also if we don't do anything in this case, the design/feature loses its claimed benefits, no need to restart Citadel to manage namespace by label/unlabeling

incfly commented:
> I would tend to say (more from a user perspective) that Citadel should handle cleaning up these key/certs automatically in disabled namespaces if possible. However, I don't fully understand the implications of doing that, and where dangerous situations which could arise, so I'll add a section in the design doc about it, and ask for feedback on it.

Customers can mis-configure stuff... It will be safer to defer the deletion until the key/cert expires. That can be a background async GC work. Yes, let's get this discussed on the design doc. Others may provide input as well.

Monkeyanator commented:
Added a watcher for namespace updates which checks if the label was added, removed, or changed. In the cases that 1) the label was removed and the default is to activate, and 2) the label was changed to "citadel.istio.io/managed=true", we generate secrets for all service accounts in the namespace. Note that `upsertSecret` handles checking for duplicate secrets, so there should not be an issue there.

incfly commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -321,13 +341,52 @@ func (sc *SecretController) scrtDeleted(obj interface{}) {
+	}
+
+	for _, sa := range serviceAccounts.Items {
+		sc.upsertSecret(sa.GetName(), sa.GetNamespace()) // idempotent, since does not gen duplicate secrets
```
If I read code correctly, call stack of this code path is initiated from namespace client go-routine, whereas, service account controller on another go-routine can hits this key/cert secret function call.

I saw your `idempotent` comment here, but several notes,

- I think this still should work, `upsertSecret` I just checked, does not modify any secret controller internal fields, so no racing on golang struct itself.
- on k8s secret object itself, I was told by  @wattli that k8s can handle concurrent modification. We observed that when customers run multiple citadel instances in a cluster. This should still work, but may add some contention that we can avoid by better coordination.
- Having said that, its may or may not worth to spending more time doing more synchronization coordination depending on how often we think the namespace flipping will happen.

Just give it a thought, don't have to put time on implementation for now.

Monkeyanator commented:
Added some logic to garbage collect certs in expiring namespaces. Each minute we check if the certificate is about to expire, and if it is, we refresh the secret: the change I made just checks if the secret is in an active namespace first, and if it isn't, deletes the secret instead.

Monkeyanator commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -410,6 +476,17 @@ func (sc *SecretController) scrtUpdated(oldObj, newObj interface{}) {
 	// one held by the ca (this may happen when the CA is restarted and
 	// a new self-signed CA cert is generated).
 	if certLifeTimeLeft < gracePeriod || !bytes.Equal(rootCertificate, scrt.Data[RootCertID]) {
+		// if the namespace is not managed, don't refresh the expired secret, delete it
```
@incfly does this seem like the right place to put secret deletion (due to deactivated namespace)?

Monkeyanator commented on security/pkg/k8s/controller/workloadsecret.go:
```diff
@@ -41,6 +41,9 @@ const (
 	IstioSecretType = "istio.io/key-and-cert"
 
+	// Label which determines whether Citadel should watch namespace
+	NamespaceManagedLabel = "citadel.istio.io/managed"
```
@incfly I'm still a bit confused on the discussion surrounding this label from the WG meeting-- on the document, we have written that this should be changed in favor of "identity.istio.io/k8s_secrets_service: namespace/uri_for_citadel".

This label seems like it's there to specify _which_ Citadel instance to activate on a namespace-- is this accurate? Does this mean that the logic for this PR will have to add another check to make sure it matches the labeled Citadel instance?

incfly replied:
My thoughts so far

- still a flag defining `manage_unlabeled_namespace` on citadel itself.
- `identity.istio.io/secret-controller-namespace: istio-system` points to the citadel namespace. Citadel does 

## Multi citadel instance in multi ns

- `ns1` runs citadel1, only responsible for ns1. `manage_unlabeled_namespace=false`, label ns1 as `identity.istio.io/secret-controller-namespace: ns1`.
-  `istio-system` runs citadel2, responsible for everything except ns1. `manage_unlabeled_namespace=true`.

## Canary use case

- `istio-system` citadel using stable version, configured as before.
- `canary-istio-system` citadel using new version, configured with manage_unlabeled_namespace=false

## Listening a list of namespace

 `manage_unlabeled_namespace=false`, label namespace `identity.istio.io/secret-controller-namespace: istio-system` for each namespace.

thoughts?

Monkeyanator replied:
What is the expected behavior when there are multiple instances of Citadel and the default for both is `manage_unlabeled_namespaces=true`? Both instances would try to operate on unlabeled namespaces (not sure if this falls under the operator's responsibilities to configure correctly or not).

incfly replied:
The behavior in that case should be not supported.

Monkeyanator replied:
Another slight concern: this feature is intended (in part) to help allow users without control plane access (multi-tenant environments) to configure where Citadel does its thing. However, that user might not know what namespace the cluster operator is even running Citadel in.

What that user would like to do say is "make sure _some_ Citadel is operating on this namespace" or "make sure _no_ Citadel is operating on this namespace". The loss of a label with a `false` value means users don't have a a good way to tell Citadel to ignore their namespace without putting a bogus tag on.

incfly commented:
Maybe one solution, to the second concern can be addressed as "
identity.istio.io/secret-controller: NONE", by having a reserved value to
the label?

On Thu, Jul 11, 2019 at 10:55 AM Sam Naser <notifications@github.com> wrote:

> *@Monkeyanator* commented on this pull request.
> ------------------------------
>
> In security/pkg/k8s/controller/workloadsecret.go
> <https://github.com/Monkeyanator/istio/pull/2#discussion_r302671081>:
>
> > @@ -41,6 +41,9 @@ const (
>  	// The Istio secret annotation type
>  	IstioSecretType = "istio.io/key-and-cert"
>
> +	// Label which determines whether Citadel should watch namespace
> +	NamespaceManagedLabel = "citadel.istio.io/managed"
>
> Another slight concern: this feature is intended (in part) to help allow
> users without control plane access (multi-tenant environments) to configure
> where Citadel does its thing. However, that user might not know what
> namespace the cluster operator is even running Citadel in.
>
> What that user would like to do say is "make sure *some* Citadel is
> operating on this namespace" or "make sure *no* Citadel is operating on
> this namespace". The loss of a label with a false value means users don't
> have a a good way to tell Citadel to ignore their namespace without putting
> a bogus tag on.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Monkeyanator/istio/pull/2?email_source=notifications&email_token=AAMVXB5ZOO44U25BIOEOVKDP65X2XA5CNFSM4H3KSUFKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6GHB7A#discussion_r302671081>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMVXB6UVGWGMWYURPM42RTP65X2XANCNFSM4H3KSUFA>
> .
>

incfly commented:
And all Citadel implementation respect that reserved value.

On Thu, Jul 11, 2019 at 10:58 AM Jianfei Hu <jianfeih@google.com> wrote:

> Maybe one solution, to the second concern can be addressed as "
> identity.istio.io/secret-controller: NONE", by having a reserved value to
> the label?
>
> On Thu, Jul 11, 2019 at 10:55 AM Sam Naser <notifications@github.com>
> wrote:
>
>> *@Monkeyanator* commented on this pull request.
>> ------------------------------
>>
>> In security/pkg/k8s/controller/workloadsecret.go
>> <https://github.com/Monkeyanator/istio/pull/2#discussion_r302671081>:
>>
>> > @@ -41,6 +41,9 @@ const (
>>  	// The Istio secret annotation type
>>  	IstioSecretType = "istio.io/key-and-cert"
>>
>> +	// Label which determines whether Citadel should watch namespace
>> +	NamespaceManagedLabel = "citadel.istio.io/managed"
>>
>> Another slight concern: this feature is intended (in part) to help allow
>> users without control plane access (multi-tenant environments) to configure
>> where Citadel does its thing. However, that user might not know what
>> namespace the cluster operator is even running Citadel in.
>>
>> What that user would like to do say is "make sure *some* Citadel is
>> operating on this namespace" or "make sure *no* Citadel is operating on
>> this namespace". The loss of a label with a false value means users
>> don't have a a good way to tell Citadel to ignore their namespace without
>> putting a bogus tag on.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Monkeyanator/istio/pull/2?email_source=notifications&email_token=AAMVXB5ZOO44U25BIOEOVKDP65X2XA5CNFSM4H3KSUFKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6GHB7A#discussion_r302671081>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAMVXB6UVGWGMWYURPM42RTP65X2XANCNFSM4H3KSUFA>
>> .
>>
>

Monkeyanator commented:
I think that's a solid idea. Pushed up a few commits implementing what we discussed here (had to use [Downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/) to figure out which NS citadel is running in)

Monkeyanator commented:
Since we discussed this at the WG meeting, should I close this PR and file a new one against istio/istio?

incfly commented:
Opening a PR against istio/istio sounds good to me.

On Fri, Jul 12, 2019 at 2:24 PM Sam Naser <notifications@github.com> wrote:

> Since we discussed this at the WG meeting, should I close this PR and file
> a new one against istio/istio?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Monkeyanator/istio/pull/2?email_source=notifications&email_token=AAMVXB5HF5BT3PS56XG4GMLP7DZATA5CNFSM4H3KSUFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ25E6A#issuecomment-511038072>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMVXB2JFOXX3XBS37UW7YDP7DZATANCNFSM4H3KSUFA>
> .
>
